fix LDAP in case of many roles on user

Enmk opened the pull request:
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Increased the default value of search_limit to 256, and added LDAP server config option to change that to an arbitrary value. Closes: #42276

Enmk committed:
fix LDAP in case of many roles on user

- Raised default value of search_limit to 256
- Added option to change that to arbitrary value
- using SipHash for computing hash of LDAP server parameters
- other minor changes

traceon commented on src/Access/ExternalAuthenticators.cpp:
```diff
@@ -184,6 +187,17 @@ void parseLDAPServer(LDAPClient::Params & params, const Poco::Util::AbstractConf
+        const auto search_limit = config.getUInt(ldap_server_config + ".search_limit");
+        // Arbitrary large number.
+        // Previously default value was 100, but in practice one client reported to have 120 roles for a user.
+        if (search_limit >= 2048)
```
`search_limit = 0` would mean no limit. So you probably want to check against that too.

However, I don't think there is any real value in introducing another limit on a manually set limit, only possibility of future problems, similar to the one we are fixing here. I thought `100` is plenty...

Enmk replied:
BTW, why do you think that having no limit (i.e. `search_limit = 0`) is not a viable option?

traceon replied:
I just followed your logic as I understood it. If no more than 2048 entries are allowed, even if the server is attempted to be configured to allow more, then unlimited number of entries should also not be allowed.

However, as I said, this additional limit is strange, and I would allow both 0 and >2048 limit to be configured by the administrator.

Enmk replied:
Removed those checks altogether, whatever integer is set in config is pushed to `params.search_limit`.

alexey-milovidov committed:
Update ExternalAuthenticators.cpp

alexey-milovidov committed:
Update LDAPClient.h

alexey-milovidov commented on src/Access/LDAPClient.cpp:
```diff
@@ -26,30 +26,30 @@ namespace ErrorCodes
-    boost::hash_combine(seed, attribute);
+    hash.update(base_dn);
+    hash.update(static_cast<int>(scope));
+    hash.update(search_filter);
```
This is wrong, as it can return identical hashes for
('a', 'bc')
and
('ab', 'c')

String length should be mixed as well.

Enmk replied:
fixed

alexey-milovidov commented on src/Common/SipHash.h:
```diff
@@ -182,6 +182,14 @@ class SipHash
+    {
+        finalize();
+        UInt128 res;
+        get128(res);
```
Calling `finalize` a second time is wrong.

alexey-milovidov replied:
Better to avoid modifying this file at all.

alexey-milovidov reviewed (changes requested):
.

Enmk committed:
Addressed issues mentioned in PR

Stronger hashing for LDAP parameters
Fixed SipHash double finalize()
Simplified checks on search_limit

Enmk commented on src/Access/LDAPClient.cpp:
```diff
@@ -26,30 +26,30 @@ namespace ErrorCodes
+    hash.update(host);
+    hash.update(port);
+    hash.update(bind_dn);
+    hash.update(user);
```
fixed as well

Enmk commented:
There was some weird issue with this PR being reported unmereable by GitHub even after merging master into topic branch several times.

alexey-milovidov committed:
Merge branch 'master' into ldap_fix_search_limit

alexey-milovidov commented:
> fatal: unable to access 'https://github.com/ClickHouse/sysroot.git/': Failed to connect to github.com port 443: Connection timed out

GitHub downtime.

alexey-milovidov committed:
Merge branch 'master' into ldap_fix_search_limit

Enmk committed:
Merge branch 'master' into ldap_fix_search_limit

Enmk committed:
Fixed build on CI/CD

Making type conversions for search_limit explicit

Enmk committed:
Merge branch 'master' into ldap_fix_search_limit

Enmk committed:
Merge branch 'master' into ldap_fix_search_limit

alexey-milovidov committed:
Merge branch 'master' into ldap_fix_search_limit

Enmk committed:
Merge branch 'master' into ldap_fix_search_limit

alexey-milovidov committed:
Merge branch 'master' into ldap_fix_search_limit

alexey-milovidov commented:
https://s3.amazonaws.com/clickhouse-test-reports/42461/0cfbf30b89f3884865d53bfae599a2eb85b08a9f/stress_test__ubsan_/bc_check_error_messages.txt

https://s3.amazonaws.com/clickhouse-test-reports/42461/0cfbf30b89f3884865d53bfae599a2eb85b08a9f/stress_test__tsan_/bc_check_error_messages.txt

alexey-milovidov commented:
@lavasser I expect it in the next release 22.12 in December.
